codec-http2/src/main/java/io/netty/handler/codec/http2/Http2FrameLogger.java

```diff
@@ -47,81 +48,121 @@ public Http2FrameLogger(InternalLogLevel level, InternalLogger logger) {
+        }
     }
 
     public void logPushPromise(Direction direction, int streamId, int promisedStreamId,
             Http2Headers headers, int padding) {
-        log(direction, "PUSH_PROMISE: streamId=%d, promisedStreamId=%d, headers=%s, padding=%d",
-                streamId, promisedStreamId, headers, padding);
+        if (enabled()) {
+            log(direction, "PUSH_PROMISE: streamId=%d, promisedStreamId=%d, headers=%s, padding=%d",
+                    streamId, promisedStreamId, headers, padding);
+        }
     }
 
     public void logGoAway(Direction direction, int lastStreamId, long errorCode, ByteBuf debugData) {
-        log(direction, "GO_AWAY: lastStreamId=%d, errorCode=%d, length=%d, bytes=%s", lastStreamId,
-                errorCode, debugData.readableBytes(), ByteBufUtil.hexDump(debugData));
+        if (enabled()) {
+            log(direction, "GO_AWAY: lastStreamId=%d, errorCode=%d, length=%d, bytes=%s", lastStreamId,
+                    errorCode, debugData.readableBytes(), toString(debugData));
+        }
     }
 
     public void logWindowsUpdate(Direction direction, int streamId, int windowSizeIncrement) {
-        log(direction, "WINDOW_UPDATE: streamId=%d, windowSizeIncrement=%d", streamId,
-                windowSizeIncrement);
+        if (enabled()) {
+            log(direction, "WINDOW_UPDATE: streamId=%d, windowSizeIncrement=%d", streamId,
+                    windowSizeIncrement);
+        }
     }
 
     public void logUnknownFrame(Direction direction, byte frameType, int streamId, Http2Flags flags, ByteBuf data) {
-        log(direction, "UNKNOWN: frameType=%d, streamId=%d, flags=%d, length=%d, bytes=%s",
-                frameType & 0xFF, streamId, flags.value(), data.readableBytes(), ByteBufUtil.hexDump(data));
+        if (enabled()) {
+            log(direction, "UNKNOWN: frameType=%d, streamId=%d, flags=%d, length=%d, bytes=%s",
+                    frameType & 0xFF, streamId, flags.value(), data.readableBytes(), toString(data));
+        }
     }
 
-    private void log(Direction direction, String format, Object... args) {
-        if (logger.isEnabled(level)) {
-            StringBuilder b = new StringBuilder(200);
-            b.append("\n----------------")
-             .append(direction.name())
-             .append("--------------------\n")
-             .append(String.format(format, args))
-             .append("\n------------------------------------");
-            logger.log(level, b.toString());
+    private boolean enabled() {
+        return logger.isEnabled(level);
+    }
+
+    private String toString(ByteBuf buf) {
+        if (level == InternalLogLevel.TRACE || buf.readableBytes() <= BUFFER_LENGTH_THRESHOLD) {
+            // Log the entire buffer.
+            return ByteBufUtil.hexDump(buf);
         }
+
+        // Otherwise just log the first 64 bytes.
+        int length = Math.min(buf.readableBytes(), BUFFER_LENGTH_THRESHOLD);
+        ByteBuf slice = buf.slice(0, length);
```

Comment:
don't have to create a slice, can use 

ByteBufUtil.hexDump(buf, buffer.readerIndex(), length)